### PR TITLE
Task 3339, rspconfig ip/netmask/gateway/vlan

### PR DESCRIPTION
**Pull request by xuweibj**
#3339 
Mini design: https://github.com/xcat2/xcat-core/wiki/rspconfig-network

UT:

* Static -> Static
```
rspconfig mid05tor12cn11 ip=172.11.139.11 netmask=255.255.0.0 gateway=0.0.0.0 vlan=11
Mon Nov  6 21:55:40 2017 mid05tor12cn11: [openbmc_debug] curl -k -c cjar -H "Content-Type: application/json" -d '{ "data": ["root", "xxxxxx"] }' https://172.11.139.222/login
Mon Nov  6 21:55:41 2017 mid05tor12cn11: [openbmc_debug] login_response 200 OK
Mon Nov  6 21:55:41 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X GET -H "Content-Type: application/json" https://172.11.139.222/xyz/openbmc_project/network/enumerate
Mon Nov  6 21:55:42 2017 mid05tor12cn11: [openbmc_debug] rspconfig_get_response 200 OK
Mon Nov  6 21:55:42 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X POST -H "Content-Type: application/json" -d '{"data":["eth0","11"]}' https://172.11.139.222/xyz/openbmc_project/network/action/VLAN
Mon Nov  6 21:55:43 2017 mid05tor12cn11: [openbmc_debug] rspconfig_vlan_response 200 OK
Mon Nov  6 21:55:45 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X POST -H "Content-Type: application/json" -d '{"data":["xyz.openbmc_project.Network.IP.Protocol.IPv4","172.11.139.11",16,"0.0.0.0"]}' https://172.11.139.222/xyz/openbmc_project/network/eth0_11/action/IP
Mon Nov  6 21:55:50 2017 mid05tor12cn11: [openbmc_debug] rspconfig_ipobject_response 200 OK
Mon Nov  6 21:55:52 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X GET -H "Content-Type: application/json" https://172.11.139.222/xyz/openbmc_project/network/enumerate
Mon Nov  6 21:55:58 2017 mid05tor12cn11: [openbmc_debug] rspconfig_check_response 200 OK
Mon Nov  6 21:55:58 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X DELETE -H "Content-Type: application/json" -d '' https://172.11.139.222/xyz/openbmc_project/network/eth0_11/ipv4/6eaf3e01
Mon Nov  6 21:55:59 2017 mid05tor12cn11: [openbmc_debug] rspconfig_delete_response 200 OK
mid05tor12cn11: BMC IP: 172.11.139.11
mid05tor12cn11: BMC Netmask: 255.255.0.0
mid05tor12cn11: BMC Gateway: 0.0.0.0
mid05tor12cn11: BMC VLAN ID: 11
```

* DHCP -> Static
```
# rspconfig mid05tor12cn11 ip=172.11.139.11 netmask=255.255.0.0 gateway=0.0.0.0 vlan=11
Mon Nov  6 21:52:56 2017 mid05tor12cn11: [openbmc_debug] curl -k -c cjar -H "Content-Type: application/json" -d '{ "data": ["root", "xxxxxx"] }' https://172.12.253.107/login
Mon Nov  6 21:52:57 2017 mid05tor12cn11: [openbmc_debug] login_response 200 OK
Mon Nov  6 21:52:57 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X GET -H "Content-Type: application/json" https://172.12.253.107/xyz/openbmc_project/network/enumerate
Mon Nov  6 21:52:57 2017 mid05tor12cn11: [openbmc_debug] rspconfig_get_response 200 OK
Mon Nov  6 21:52:57 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X POST -H "Content-Type: application/json" -d '{"data":["eth0","11"]}' https://172.12.253.107/xyz/openbmc_project/network/action/VLAN
Mon Nov  6 21:52:59 2017 mid05tor12cn11: [openbmc_debug] rspconfig_vlan_response 200 OK
Mon Nov  6 21:53:01 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X POST -H "Content-Type: application/json" -d '{"data":["xyz.openbmc_project.Network.IP.Protocol.IPv4","172.11.139.11",16,"0.0.0.0"]}' https://172.12.253.107/xyz/openbmc_project/network/eth0_11/action/IP
Mon Nov  6 21:53:08 2017 mid05tor12cn11: [openbmc_debug] rspconfig_ipobject_response 200 OK
Mon Nov  6 21:53:10 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X GET -H "Content-Type: application/json" https://172.12.253.107/xyz/openbmc_project/network/enumerate
Mon Nov  6 21:53:16 2017 mid05tor12cn11: [openbmc_debug] rspconfig_check_response 200 OK
Mon Nov  6 21:53:16 2017 mid05tor12cn11: [openbmc_debug] curl -k -b cjar -X PUT -H "Content-Type: application/json" -d '{"data":0}' https://172.12.253.107/xyz/openbmc_project/network/eth0/attr/DHCPEnabled
Mon Nov  6 21:53:18 2017 mid05tor12cn11: [openbmc_debug] rspconfig_dhcpdis_response 200 OK
mid05tor12cn11: BMC IP: 172.11.139.11
mid05tor12cn11: BMC Netmask: 255.255.0.0
mid05tor12cn11: BMC Gateway: 0.0.0.0
mid05tor12cn11: BMC VLAN ID: 11
```

Error handle:
```
# rspconfig mid05tor12cn11 ip=172.11.139.11 netmask=255.255.0.0
mid05tor12cn11: Error: Please config IP netmask gateway at the same time
# rspconfig mid05tor12cn11 ip=172.11.139.11 netmask=255.255.0.0 gateway=1.1.
mid05tor12cn11: Error: Invalid parameter for option gateway: 1.1.
```

To workaround issue https://github.com/openbmc/openbmc/issues/2556, after setting VLAN or adding new object, will sleep 3 seconds.

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 1]
```diff
@@ -748,40 +789,55 @@ sub parse_args {
             return ([ 1, "Unsupported command: $command $subcommand" ]);
         }
     } elsif ($command eq "rspconfig") {
+        my $num_subcommand = @ARGV;
         my $setorget;
+        my $all_subcommand = "";
         foreach $subcommand (@ARGV) {
             if ($subcommand =~ /^(\w+)=(.*)/) {
                 return ([ 1, "Can not configure and display nodes' value at the same time" ]) if ($setorget and $setorget eq "get");
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Can we change this to: `Can not configure and display nodes at the same time.`

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 2]
```diff
@@ -748,40 +789,55 @@ sub parse_args {
...
                 my $key = $1;
                 my $value = $2;
                 return ([ 1, "Changing ipsrc value is currently not supported." ]) if ($key eq "ipsrc");
                 return ([ 1, "Unsupported command: $command $key" ]) unless ($key =~ /^ip$|^netmask$|^gateway$|^hostname$|^vlan$/);
+                return ([ 1, "Can not configure hostname and other options at the same time" ]) if ($key eq "hostname" and $num_subcommand > 1);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: @gurevichmark Any comment on this line or OK?  Debating something like.... 

* Can not set hostname alongside other options. 
* hostname must be set as a single command.

Or just leave it as is? 

Can we add a `.` at the end?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 3]
```diff
@@ -748,40 +789,55 @@ sub parse_args {
...
-                    if ($value ne "dhcp" and !xCAT::NetworkUtils->isIpaddr($value)) {
-                        return ([ 1, "Invalid parameter for option $key: $value" ]);
+                    if ($value ne "dhcp" ) {
+                        if (!xCAT::NetworkUtils->isIpaddr($value)) {
+                            return ([ 1, "Invalid parameter for option $key: $value" ]);
+                        } else {
+                            $all_subcommand .= $key . ",";
+                        }
+                    } else {
+                        return ([ 1, "Can not configure IP to DHCP and other options at the same time" ]) if ($num_subcommand > 1);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Similar to above...   `Setting ip=dhcp must be issued without other options.`   OR maybe it's ok...

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 4]
```diff
@@ -748,40 +789,55 @@ sub parse_args {
...
-                #
-                unless (($key eq "ip" and $value eq "dhcp") or $key eq "hostname") {
-                    $check = unsupported($callback); if (ref($check) eq "ARRAY") { return $check; }
-                }
             } elsif ($subcommand =~ /^ip$|^netmask$|^gateway$|^hostname$|^vlan$|^ipsrc$/) {
                 return ([ 1, "Can not configure and display nodes' value at the same time" ]) if ($setorget and $setorget eq "set");
                 $setorget = "get";
             } elsif ($subcommand =~ /^sshcfg$/) {
-                $setorget = ""; # SSH Keys are copied using a RShellAPI, not REST API
+                return ([ 1, "Can not configure sshcfg and other options at the same time" ]) if ($num_subcommand > 1);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Add a `.` to the end?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 5]
```diff
@@ -748,40 +789,55 @@ sub parse_args {
...
             } else {
                 return ([ 1, "Unsupported command: $command $subcommand" ]);
             }
-        }  
+        }
+        if ($all_subcommand) {
+            if ($all_subcommand !~ /ip/ or $all_subcommand !~ /netmask/ or $all_subcommand !~ /gateway/) {
+                if ($all_subcommand =~ /vlan/) {
+                    return ([ 1, "Please config IP netmask gateway with VLAN" ]);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Suggest:  VLAN must be configured with IP, netmask, and gateway

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 6]
```diff
@@ -748,40 +789,55 @@ sub parse_args {
...
+                } else {
+                    return ([ 1, "Please config IP netmask gateway at the same time" ]);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Suggest: IP, netmask, and gateway must be configured together.

**Comment by whowutwut** [comment]
@xuweibj So it looks like we are not going to allow a single command to be modified by getting the configuration today and creating a new object overlaying the values provided?

**Comment by xuweibj** [comment]
@whowutwut , yes, we could just  creating a new object now. I think if necessary we could make it better in the future.

**Comment by xuweibj** [comment]
Will open a new one.

**Pull request closed without merging**